utilix/rundb.py

```diff
@@ -153,46 +172,52 @@ def new_token(self):
         path = PREFIX + "/login"
         username = uconfig.get('RunDB', 'rundb_api_user')
         pw = uconfig.get('RunDB', 'rundb_api_password')
-        data=json.dumps({"username": username,
-                         "password": pw})
+        data = json.dumps({"username": username,
+                           "password": pw})
+        logger.debug('Creating a new token: doing API call now')
         response = requests.post(path, data=data, headers=BASE_HEADERS)
-        return json.loads(response.text)['access_token'], username
+        response_json = json.loads(response.text)
+        logger.debug(f'The response contains these keys: {list(response_json.keys())}')
+        token = response_json.get('access_token', 'CALL_FAILED')
+        if token == 'CALL_FAILED':
+            logging.error(f"API call to create new token failed. Here is the response:\n{response.text}")
+            raise RuntimeError("Creating a new token failed.")
+        self.string = token
+        self.user = username
+        self.creation_time = datetime.datetime.now().timestamp()
+        self.write()
 
     @property
     def is_valid(self):
         # TODO do an API call for this instead?
-        return datetime.datetime.now().timestamp() - self.creation_time < 24*60*60
-
-    def refresh(self):
-        # check if user in xenon_config matches the token...
-        # if it doesn't, use the one in xenon_config and overwrite the token
-        if self.user != uconfig.get('RunDB', 'rundb_api_user'):
-            logger.debug(f"Username in {uconfig.config_path} does not match token. Overwriting the token.")
-            self.string, self.user = self.new_token()
-
+        diff = datetime.datetime.now().timestamp() - self.creation_time - 100
+        return diff < 24*60*60
 
-        # if valid, don't do anything
-        if self.is_valid:
-            logger.debug("Token is valid")
-            return
+    @property
+    def json(self):
+        return dict(string=self.string, creation_time=self.creation_time, user=self.user)
 
+    def refresh(self):
         # update the token string
         url = PREFIX + "/refresh"
         headers = BASE_HEADERS.copy()
-        headers['Authorization'] = "Bearer {string}".format(string=self.string)
+        headers['Authorization'] = f"Bearer {self.string}"
+        logger.debug(f"Refreshing your token with API call {url}")
         response = requests.get(url, headers=headers)
-        response = json.loads(response.text)
-
-        # if rewew fails, try logging back in
-        if (response['status_code'] is not 200 and response['error'] != 'EarlyRefreshError'):
-            print("Refreshing token")
-            self.string, self.user = self.new_token()
+        logger.debug(f"The response was {response.text}")
+        # if renew fails, try logging back in
+        if response.status_code is not 200:
```

Comment:
Maybe also up this syntax:
![afbeelding](https://user-images.githubusercontent.com/22295914/101142645-2b62ae80-3616-11eb-9e66-512e88384be4.png)

Reply:
done